### PR TITLE
p2p: adding peerstore

### PR DESCRIPTION
**Pull request by shiqizng**
This PR adds a peerstore module.

### CODE REVIEW

File: network/peerstore/peerstore.go
Comment (shiqizng) [inline network/peerstore/peerstore.go:36]: using leveldb kv store since pebbleDB is labeled experimental.

File: network/peerstore/peerstore_test.go
Comment (tzaffi) [inline network/peerstore/peerstore_test.go:72]: I recommend adding some error cases as well here or in a separate test. What happens when the addrs aren't formatted properly, or have invalid peerID, or invalid ip4, or adhere to our old format? Some of these would be useful assertions.

File: network/peerstore/peerstore.go
Comment (tzaffi) [inline network/peerstore/peerstore.go:50]: Similar suggestion about wrapping the error

File: network/peerstore/peerstore.go
Comment (tzaffi) [inline network/peerstore/peerstore.go:62]: ditto

File: network/peerstore/peerstore.go
Comment (tzaffi) [inline network/peerstore/peerstore.go:74]: ditto

File: network/peerstore/peerstore.go
Comment (tzaffi) [inline network/peerstore/peerstore.go:78]: ditto

**Review by tzaffi — COMMENTED** [review]
Looks very promising and clean!

I made some suggestions regarding adding test cases and wrapping errors from external libraries with our own messages to improve debuggability.

File: network/peerstore/peerstore.go
Comment (Eric-Warehime) [reply network/peerstore/peerstore.go:36]: Maybe in a future PR? As discussed, once we understand more about how the i/o usage of the peerstore is working we might want to test out different backends. We should take a look at what "experimental" means for the pebble implementation because it seems like most of our kv backed stuff is using that. Maybe there's something we can contribute upstream there as well?

File: network/peerstore/peerstore.go
Comment (Eric-Warehime) [inline network/peerstore/peerstore.go:43]: Maybe we want to accept `[]peer.PeerInfo` here and move the internals of this to util functions. Seems like converting between string, Multiaddr, and PeerInfo is going to be used across a few places.

File: network/peerstore/peerstore.go
Comment (shiqizng) [reply network/peerstore/peerstore.go:36]: experimental probably means these work are still in progress, https://github.com/ipfs/go-ds-pebble#status

Comment (Eric-Warehime) [reply network/peerstore/peerstore.go]: Yeah out of that list:

> The road to maturity includes:
> * Benchmarks against go-ds-badger, go-ds-badger2, and go-ds-leveldb.
> * Exposing metrics.
> * Testing it in IPFS, libp2p, and Lotus, and characterising its behaviour, performance and footprint under real, practical workloads.

Two of those are testing and benchmarking--things which we plan to do during our integration. And the other is a feature request that maybe we could upstream if we ended up needing. So maybe it's still worth us looking at using it.

Comment (iansuvak) [reply network/peerstore/peerstore.go]: Yeah levelDB would be 3rd DB tech. I thought that we mentioned using Sqlite? I know it's not what's used by the maintainers but is stable and is something we are using already.

On the other hand we will be using pebbleDB as @Eric-Warehime mentioned.

I prefer either of the two to levelDB for those two reasons

File: network/p2p/peerstore/utils.go
Comment (iansuvak) [inline network/p2p/peerstore/utils.go:30]: So this will hard fail if a single input string of a potentially longish list is malformed. Is this the behavior that we want or should we log the error and continue?

The use-cases I'm thinking of here are if we use this codepath for `-p` option and phonebooks that users could manually edit. A typo could potentially prevent the node from accepting any of the inputs.

File: network/peerstore/peerstore.go
Comment (shiqizng) [reply network/peerstore/peerstore.go]: changed to use pebbledb.

File: network/p2p/peerstore/utils.go
Comment (shiqizng) [reply network/p2p/peerstore/utils.go:30]: good point. I will also return a list of bad addresses.

File: network/p2p/peerstore/peerstore.go
Comment (Eric-Warehime) [inline network/p2p/peerstore/peerstore.go:56]:
```suggestion
		ps.AddAddrs(info.ID, info.Addrs, libp2p.AddressTTL)
```

I'm not sure we want the initial set of peers to be permanent. We'll need some other way of specifying that a group of peers is intended for permanent peering--and a separate way of referring to that group--in a future PR.

File: network/p2p/peerstore/peerstore.go
Comment (tzaffi) [inline network/p2p/peerstore/peerstore.go:24]: `go-ds-pebble` [pulls in the ipfs logger](https://github.com/ipfs/go-ds-pebble/blob/master/datastore.go#L16). Is this a problem?

Comment (tzaffi) [reply network/p2p/peerstore/peerstore.go:24]: Answering my own question. This is a wrapper around Uber's zap and we already have some references in the go.sum:

image

File: network/p2p/peerstore/peerstore.go
Comment (tzaffi) [inline network/p2p/peerstore/peerstore.go:46]: ignoreable nit: "invalid path" seems like the most likely culprit, but maybe there are other reasons pebble could error on it's `Open()`? Also, is it safe to print out the failing path as well?

**Review by tzaffi — DISMISSED** [review]
My previous questions have been addressed and new questions don't block.

File: network/p2p/peerstore/peerstore.go
Comment (shiqizng) [reply network/p2p/peerstore/peerstore.go]: updated.

**Review by Eric-Warehime — APPROVED** [review]
I'm happy to merge this now. We can adjust any things which end up needing changes in any future PRs as necessary.

**Review by winder — APPROVED** [review]

**Review by iansuvak — APPROVED** [review]

**Pull request merged**
